### PR TITLE
`General`: Allow admins to hide the password-based login form

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

@LDAP and @dfuchss  at KIT aim to completely disable password login and rely only on SAML2. Having a password login causes confusion since Artemis credentials do not match SAML2 credentials.
When both authentication options were enabled, it is not obvious to users that when using Single Sign-on they do not have to put anything into the login form.

### Description

A new configuration option `info.saml2.password-login-disabled` was added to the SAML2 config file.
I implemented the client changes, so that this configuration option disables the traditional login form. I also designed the user interface to now make it clear that there are two options to sign in. The UI is responsive and adjusts itself based on the admin's configuration.
The query parameter `?showLoginForm` overrides the configuration and displays the password-based login nevertheless (useful for admin login).
Furthermore, I added the `info.saml2.idenitity-provider-name` configuration setting to optionally add the name of the identity provider to the title on the sign-in page.

### Steps for Testing

Prerequisites:
- 1 Administrator

1. Start Artemis locally with SAML2 authentication enabled
2. Make sure both sign on options are visible, working and UI is responsive
3. Set `info.saml2.password-login-disabled: true` in your local configuration
4. Ensure only SSO authentication is visible
5. Adding the query parameter`?showLoginForm` to the URL should also display the traditional login form (override)
6. Disable SAML2 authentication, the password form should appear even though `password-login-disabled` is set (fallback)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Both traditional and SSO authentication enabled:

Screenshot 2022-06-26 at 16-35-24 Welcome to Artemis!

Only password-based sign on:

Screenshot 2022-06-26 at 16-17-40 Welcome to Artemis!

Only SSO (e.g., SAML2) authentication:

Screenshot 2022-06-26 at 16-16-51 Welcome to Artemis!

### CODE REVIEW

File: src/main/resources/config/application-saml2.yml [hunk 1]
```diff
@@ -47,6 +47,8 @@ saml2:
         #   registrationid: <id>
         #   entityid: <id>
 
+# The name of the SAML2 identity provider (optional)
+info.saml2.idenitity-provider-name:
```
Comment (dfuchss) [inline src/main/resources/config/application-saml2.yml:51]:
```suggestion
info.saml2.identity-provider-name:
```

**Review by dfuchss — APPROVED** [review]
@ge65cer nice work. I would suggest to give it into the review. I've tested it on our test systems. Works as expected :)

image

File: src/main/webapp/app/home/home.component.ts [hunk 2]
```diff
@@ -136,23 +136,15 @@ export class HomeComponent implements OnInit, AfterViewChecked {
...
-        } else if (this.saml2Config) {
-            const btn = this.renderer.selectRootElement('#saml2Button', true);
-            if (btn) {
-                btn.focus();
-                this.inputFocused = true;
-            }
+        const mainElement = this.renderer.selectRootElement(this.passwordLoginDisabled ? '#saml2Button' : '#username', true);
+        if (mainElement) {
+            mainElement.focus();
+            this.mainElementFocused = true;
```
Comment (dfuchss) [inline src/main/webapp/app/home/home.component.ts:147]: I think here was a small semantic change. If passwordLoginDisabled == true and saml2Config == false

**Comment by sleiss** [comment]
I am really looking forward to this feature, I've implemented something very similar for the Hochschule München-Setup.
One small idea: Do you think it would make sense to offer a way for admins to use a local login by e.g. specifying a query parameter?
Gitlab offers a similar feature: https://docs.gitlab.com/ee/integration/saml.html#auto_sign_in_with_provider (`?auto_sign_in=false`).

File: src/main/resources/config/application.yml [hunk 3]
```diff
@@ -219,6 +219,9 @@ info:
     # default value set to false
     text-assessment-analytics-enabled: false
     browser-fingerprints-enabled: true
+    # Disables the password-based login user interface, but leaves the API enabled.
+    # Can be useful with alternative login flows (e.g., SAML2).
+    password-login-disabled: false
```
Comment (krusche) [inline src/main/resources/config/application.yml:224]: can we move this into a different section? I think it should be part of the section where the SAML2 related information is stored

File: src/main/webapp/app/home/home.component.ts (hunk 2 above)
Comment (ge65cer) [reply src/main/webapp/app/home/home.component.ts:147]: Yes, I am aware of that. The `else if` was not needed as the variable is defined [like so](https://github.com/ls1intum/Artemis/pull/5291/files#diff-c3974a25b9c2e1f16a1c5673271c038fe8f54e66af898a68c131904b4a43f11dR123).

File: src/main/resources/config/application.yml (hunk 3 above)
Comment (ge65cer) [reply src/main/resources/config/application.yml:224]: Sure, I will move it to the SAML2 config (`info.saml2.password-login-disabled`).

**Comment by dfuchss** [comment]
> One small idea: Do you think it would make sense to offer a way for admins to use a local login by e.g. specifying a query parameter?

I also thought about a similar feature for a follow up PR. For a first implementation I would skip that. But we should track this idea in a Feature (Issue) as it would be very useful :)

**Comment by ge65cer** [comment]
> > One small idea: Do you think it would make sense to offer a way for admins to use a local login by e.g. specifying a query parameter?
> 
> I also thought about a similar feature for a follow up PR. For a first implementation I would skip that. But we should track this idea in a Feature (Issue) as it would be very useful :)

I don't think this would be too complicated. Essentially you want a query parameter to override the `password-login-disabled` option, right? E.g., `?showLoginForm` would display the login form nevertheless?

**Comment by dfuchss** [comment]
> > > One small idea: Do you think it would make sense to offer a way for admins to use a local login by e.g. specifying a query parameter?
> > 
> > I also thought about a similar feature for a follow up PR. For a first implementation I would skip that. But we should track this idea in a Feature (Issue) as it would be very useful :)
> 
> I don't think this would be too complicated. Essentially you want a query parameter to override the `password-login-disabled` option, right? E.g., `?showLoginForm` would display the password-based login nevertheless?

If it's easily possible I would opt for including the feature in this PR :) If not in a Follow-up PR :)

**Comment by sleiss** [comment]
> > > One small idea: Do you think it would make sense to offer a way for admins to use a local login by e.g. specifying a query parameter?
> > 
> > 
> > I also thought about a similar feature for a follow up PR. For a first implementation I would skip that. But we should track this idea in a Feature (Issue) as it would be very useful :)
> 
> I don't think this would be too complicated. Essentially you want a query parameter to override the `password-login-disabled` option, right? E.g., `?showLoginForm` would display the password-based login nevertheless?

Exactly, that's what I would suggest.

**Review by b-fein — APPROVED** [review]
Tested manually: all combinations of en-/disabling SAML2/passwword-logins worked as expected and described. The divider in the UI even switches to a vertical layout when e.g., on mobile. :+1: 

Code changes look good, too.

**Comment by ge65cer** [comment]
@sleiss @dfuchss Implemented the query parameter override.

**Review by dfuchss — APPROVED** [review]
Code changes look good

**Review by b-fein — DISMISSED** [review]
Re-approve code and manual testing.

When using the URL parameter, the username/password login form is shown.

**Review by willergeorg — DISMISSED** [review]
Client code lgtm. 👍

File: src/main/webapp/app/home/home.component.ts [hunk 4]
```diff
@@ -39,8 +40,11 @@ export class HomeComponent implements OnInit, AfterViewChecked {
     captchaRequired = false;
     credentials: Credentials;
     isRegistrationEnabled = false;
+    passwordLoginDisabled = false;
+    loginFormOverride = false;
+    saml2Config: Saml2Config | undefined;
```
Comment (krusche) [inline src/main/webapp/app/home/home.component.ts:45]: we wanted to keep the home component independent of Saml2Config.
Is it really needed?

File: src/main/webapp/app/home/home.component.ts [hunk 5]
```diff
@@ -39,8 +40,11 @@ export class HomeComponent implements OnInit, AfterViewChecked {
     captchaRequired = false;
     credentials: Credentials;
     isRegistrationEnabled = false;
+    passwordLoginDisabled = false;
+    loginFormOverride = false;
```
Comment (krusche) [inline src/main/webapp/app/home/home.component.ts:44]: Why do we need two boolean values here?

**Review by b-fein — APPROVED** [review]

**Review by mallgaier — APPROVED** [review]
Code changes look good to me 👍

**Review by LDAP — APPROVED** [review]

**Review by willergeorg — APPROVED** [review]
Code changes look reasonable. 👍

**Review by Dominik-Weinzierl — APPROVED** [review]
Tested the layout on my mobile phone - looks good :+1:

**Review by willergeorg — APPROVED** [review]
Tested on the mobile device during the testing session. Looks as expected. 👍

**Review by akesfeden — APPROVED** [review]
Tested on mobile device during testing session, works as expected.

**Review by babyygemperor — APPROVED** [review]
Tested on mobile device, looks good

**Review by nairaarnez — APPROVED** [review]
Approving after testing on movile device, points 2, 4, 5 and 6 work correctly

**Review by bassner — APPROVED** [review]

**Comment by krusche** [comment]
e2e tests are failing at the moment. I triggered them again, let's see if they pass.
If not, we should investigate what's going on. They seem to fail constantly for this PR and passed in all other PRs that I recently merged.

**Pull request merged**